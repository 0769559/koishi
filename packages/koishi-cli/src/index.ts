import * as commonPlugin from 'koishi-plugin-common'
import * as schedulePlugin from 'koishi-plugin-schedule'
export { commonPlugin, schedulePlugin }

export * from 'koishi-core'
export * from 'koishi-utils'
export * from 'koishi-plugin-schedule/dist/utils'
export { AppConfig, PluginConfig } from './worker'

const version = require('../package').version as string
export { version }
