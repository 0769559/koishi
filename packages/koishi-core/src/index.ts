export * from './app'
export * from './command'
export * from './context'
export * from './database'
export * from './meta'
export * from './sender'
export * from './server'
export * from './plugins/help'
export * from './plugins/suggest'
export * from './plugins/throttle'
export * from './plugins/queue'
export * from './plugins/validate'

import './plugins/help'
import './plugins/suggest'
import './plugins/throttle'
import './plugins/queue'
import './plugins/validate'
