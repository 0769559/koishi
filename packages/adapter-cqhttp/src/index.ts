/* eslint-disable quote-props */

import { Server, Context, Session } from 'koishi-core'
import HttpServer from './http'
import WsClient from './ws'
import WsServer from './ws-reverse'

export * from './api'

declare module 'koishi-core/dist/server' {
  interface BotOptions {
    token?: string
  }

  interface ServerTypes {
    'http': typeof HttpServer
    'ws': typeof WsClient
    'ws-reverse': typeof WsServer
  }
}

Server.types['http'] = HttpServer
Server.types['ws'] = WsClient
Server.types['ws-reverse'] = WsServer

Session.prototype.$send = async function $send (this: Session, message: string, autoEscape = false) {
  if (!message) return
  let ctxId: number
  // eslint-disable-next-line no-cond-assign
  const ctxType = (ctxId = this.groupId) ? 'group' : (ctxId = this.userId) ? 'user' : null
  if (this.$app.options.preferSync) {
    ctxType === 'group'
      ? await this.$bot.sendGroupMsg(ctxId, message, autoEscape)
      : await this.$bot.sendPrivateMsg(ctxId, message, autoEscape)
    return
  }
  if (this.$response) {
    const session = this.$bot.createSession(this.messageType, ctxType, ctxId, message)
    if (this.$app.bail(this, 'before-send', session)) return
    return this.$response({ reply: message, autoEscape, atSender: false })
  }
  return ctxType === 'group'
    ? this.$bot.sendGroupMsgAsync(ctxId, message, autoEscape)
    : this.$bot.sendPrivateMsgAsync(ctxId, message, autoEscape)
}
