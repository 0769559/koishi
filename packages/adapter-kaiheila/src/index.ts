import { App, Server } from 'koishi-core'
import { Logger, defineProperty, assertProperty } from 'koishi-utils'
import { createSession, KaiheilaBot } from './bot'
import { createHmac } from 'crypto'
import axios from 'axios'

interface KaiheilaOptions {
  path?: string
}

declare module 'koishi-core/dist/app' {
  interface AppOptions {
    kaiheila?: KaiheilaOptions
  }
}

const logger = new Logger('server')

export default class HttpServer extends Server<KaiheilaBot> {
  constructor(app: App) {
    assertProperty(app.options, 'port')
    super(app, KaiheilaBot)
  }

  private async _listen(bot: KaiheilaBot) {
    bot.ready = true
  }

  async listen() {
    const { kaiheila = {} } = this.app.options
    const { path = '/' } = kaiheila
    this.app.router.post(path, (ctx) => {
      logger.debug('receive %o', ctx.request.body)
      const session = createSession(this.app, ctx.request.body)
      ctx.body = {}

      // dispatch events
      console.log(session)
      // this.dispatch(session)
    })

    await Promise.all(this.bots.map(bot => this._listen(bot)))
  }

  close() {
    logger.debug('http server closing')
  }
}
