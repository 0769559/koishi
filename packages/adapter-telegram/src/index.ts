import { Server, Context } from 'koishi-core'
import HttpServer from './http'
import axios from 'axios'

interface TelegramOptions {
  url?: string
  selfUrl?: string
  preferSync?: boolean
}
declare module 'koishi-core/dist/server' {
  interface BotOptions {
    token?: string
  }
}
declare module 'koishi-core/dist/app' {
  interface AppOptions {
    telegram?: TelegramOptions
  }
}
declare module 'koishi-core/dist/session' {
  namespace Session {
    interface Events {
      'lifecycle': {
        'enable': {}
        'disable': {}
        'connect': {}
        'heartbeat': {}
      }
    }
  }
}

export * from './bot'
export * from './http'

Server.types.telegram = HttpServer
const { broadcast } = Context.prototype
const imageRE = /\[CQ:image,file=([^,]+),url=([^\]]+)\]/

Context.prototype.broadcast = async function boardcast(this: Context, ...args: any[]) {
  const index = Array.isArray(args[0]) ? 1 : 0
  let message = args[index] as string
  let output = ''
  let capture: RegExpExecArray
  // eslint-disable-next-line no-cond-assign
  while (capture = imageRE.exec(message)) {
    const [text, , url] = capture
    output += message.slice(0, capture.index)
    message = message.slice(capture.index + text.length)
    const { data } = await axios.get<ArrayBuffer>(url, { responseType: 'arraybuffer' })
    output += `[CQ:image,file=base64://${Buffer.from(data).toString('base64')}]`
  }
  args[index] = output + message
  return broadcast.apply(this, args)
}
