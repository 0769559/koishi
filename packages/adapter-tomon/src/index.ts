import Tomon from 'tomon-sdk'
import { App, Server, Session } from 'koishi-core'
import { TomonBot } from './bot'
import { camelize } from 'koishi-utils'

export * from './bot'

declare module 'koishi-core/dist/server' {
  interface BotOptions {
    token?: string
  }
}

Server.types.tomon = class TomonServer extends Server<TomonBot> {
  constructor(app: App) {
    super(app, TomonBot)
  }

  async _listen(bot: TomonBot) {
    const tomon = bot.tomon = new Tomon()
    await tomon.start(bot.token)
    bot.ready = true
    const selfId = bot.selfId = tomon.discriminator
    this.app.bots[selfId] = bot

    tomon.on('MESSAGE_CREATE', async ({ d }) => {
      const userId = d.author.discriminator
      if (userId === selfId) return
      this.dispatch(new Session(this.app, {
        ...camelize(d),
        selfId,
        userId,
        kind: 'tomon',
        message: d.content || '', // TODO 处理表情包和图片
        postType: 'message',
        messageType: d['guild_id'] ? 'group' : 'private',
        groupId: d['guild_id'],
      }))
    })

    return new Promise<void>((resolve) => {
      tomon.on('NETWORK_CONNECTED', resolve)
    })
  }

  async listen() {
    await Promise.all(this.bots.map(bot => this._listen(bot)))
  }

  close() {}
}
