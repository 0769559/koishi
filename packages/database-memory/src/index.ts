import { injectMethods, TableType, TableData, createUser, User, createGroup, getSelfIds, Group, InjectConfig, registerDatabase } from 'koishi-core'
import { observe } from 'koishi-utils'

declare module 'koishi-core/dist/database' {
  interface Subdatabases {
    memory: MemoryDatabase
  }

  interface DatabaseConfig {
    memory?: MemoryConfig
  }
}

interface MemoryConfig {}

export class MemoryDatabase {
  store: { [T in TableType]?: TableData[T][] } = {}

  constructor (public config: MemoryConfig, public injectConfig: InjectConfig<'memory'>) {}

  start () {
    for (const key in this.injectConfig) {
      this.store[key] = []
    }
  }

  async create <K extends TableType> (table: K, data: Partial<TableData[K]>) {
    const store = this.store[table]
    if (typeof data.id !== 'number') {
      let index = 1
      while (index in store) index++
      data.id = index
    }
    return store[data.id] = data as TableData[K]
  }

  async remove <K extends TableType> (table: K, id: number) {
    delete this.store[table][id]
  }

  async update <K extends TableType> (table: K, id: number, data: Partial<TableData[K]>) {
    Object.assign(this.store[table][id], clone(data))
  }

  async count (table: TableType) {
    return Object.keys(this.store[table]).length
  }
}

registerDatabase('memory', MemoryDatabase)

function clone <T> (source: T): T {
  return JSON.parse(JSON.stringify(source))
}

injectMethods('memory', 'user', {
  async getUser (userId, authority) {
    authority = typeof authority === 'number' ? authority : 0
    const data = this.store.user[userId]
    if (data) return clone(data)
    if (authority < 0) return null
    const fallback = createUser(userId, authority)
    if (authority) this.store.user[userId] = fallback
    return clone(fallback)
  },

  async getUsers (...args) {
    if (args.length > 1 || args.length && typeof args[0][0] !== 'string') {
      return Object.keys(this.store.user)
        .filter(id => args[0].includes(+id))
        .map(id => clone(this.store.user[id]))
    } else {
      return Object.values(this.store.user)
    }
  },

  async setUser (userId, data) {
    return this.update('user', userId, data)
  },

  async observeUser (user, authority) {
    if (typeof user === 'number') {
      const data = await this.getUser(user, authority)
      return data && observe(clone(data), diff => this.setUser(user, diff), `user ${user}`)
    }

    const data = await this.getUser(user.id, authority)
    if ('_diff' in user) return (user as User)._merge(clone(data))
    return observe(Object.assign(user, clone(data)), diff => this.setUser(user.id, diff), `user ${user.id}`)
  },
})

injectMethods('memory', 'group', {
  async getGroup (groupId, selfId) {
    selfId = typeof selfId === 'number' ? selfId : 0
    const data = this.store.group[groupId]
    if (data) return clone(data)
    const fallback = createGroup(groupId, selfId)
    if (selfId && groupId) this.store.group[groupId] = fallback
    return clone(fallback)
  },

  async getAllGroups (...args) {
    const assignees = args.length > 1 ? args[1]
      : args.length && typeof args[0][0] === 'number' ? args[0] as never
        : await getSelfIds()
    if (!assignees.length) return []
    return Object.keys(this.store.group)
      .filter(id => assignees.includes(this.store.group[id].assignee))
      .map(id => clone(this.store.group[id]))
  },

  async setGroup (groupId, data) {
    return this.update('group', groupId, data)
  },

  async observeGroup (group, selfId) {
    if (typeof group === 'number') {
      const data = await this.getGroup(group, selfId)
      return data && observe(clone(data), diff => this.setGroup(group, diff), `group ${group}`)
    }

    const data = await this.getGroup(group.id, selfId)
    if ('_diff' in group) return (group as Group)._merge(clone(data))
    return observe(Object.assign(group, clone(data)), diff => this.setGroup(group.id, diff), `group ${group.id}`)
  },
})
