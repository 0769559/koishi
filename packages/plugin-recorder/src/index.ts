import { WriteStream, createWriteStream, existsSync, mkdirSync } from 'fs'
import { resolve, dirname } from 'path'
import { Meta, Context } from 'koishi-core'

declare module 'koishi-core/dist/context' {
  interface EventMap {
    'before-record' (meta: Meta): any
  }
}

function pick <T, K extends keyof T> (source: T, keys: K[]): Pick<T, K> {
  return keys.reduce((prev, curr) => (prev[curr] = source[curr], prev), {} as Pick<T, K>)
}

const streams: Record<string, WriteStream> = {}

export interface RecorderOptions {
  folder?: string
}

const cwd = process.cwd()

export const name = 'recorder'

export function apply (ctx: Context, options: RecorderOptions = {}) {
  async function handleMessage (meta: Meta<never, 'assignee'>) {
    if (meta.$ctxType !== 'group' || meta.postType === 'message' && meta.$group.assignee !== ctx.app.selfId) return
    if (await ctx.serialize('before-record', meta)) return
    const output = JSON.stringify(pick(meta, ['time', 'userId', 'message'])) + '\n'
    const path = resolve(cwd, options.folder || 'messages', `${meta.groupId}.txt`)
    if (!streams[path]) {
      const folder = dirname(path)
      if (!existsSync(folder)) {
        mkdirSync(folder, { recursive: true })
      }
      streams[path] = createWriteStream(path, { flags: 'a' })
    }
    streams[path].write(output)
  }

  ctx.on('attach-group', (meta: Meta<never, 'assignee'>) => {
    handleMessage(meta)
  })

  ctx.on('before-send', (meta: Meta<never, 'assignee'>) => {
    handleMessage(meta)
  })
}
