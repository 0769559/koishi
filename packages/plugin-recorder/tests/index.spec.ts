import { MockedApp, BASE_SELF_ID } from 'koishi-test-utils'
import { startAll, stopAll } from 'koishi-core'
import { readFileSync } from 'fs-extra'
import { resolve } from 'path'
import { sleep } from 'koishi-utils'
import * as recorder from '../src'
import del from 'del'

const app1 = new MockedApp()
const app2 = new MockedApp({ selfId: BASE_SELF_ID + 1 })

app1.plugin(recorder)
app2.plugin(recorder)

beforeAll(() => startAll())

const outFolder = resolve(process.cwd(), 'messages')

afterAll(() => del(outFolder))

test('group message', async () => {
  const mock = jest.fn()
  app1.receiver.on('record-writing', mock)
  await app1.receiveMessage('group', 'foo', 123, 321)
  await app1.receiveMessage('group', 'bar', 456, 654)
  await app1.receiveMessage('group', 'baz', 789, 321)
  expect(mock).toBeCalledTimes(3)
})

test('private message', async () => {
  const mock = jest.fn()
  app1.receiver.on('record-writing', mock)
  await app1.receiveMessage('user', 'foo', 123)
  expect(mock).toBeCalledTimes(0)
})

test('check content', async () => {
  await stopAll()
  await sleep(100)
  expect(readFileSync(resolve(outFolder, '321.txt'), 'utf8')).toMatchSnapshot()
})
