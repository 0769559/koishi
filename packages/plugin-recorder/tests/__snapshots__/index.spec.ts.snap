// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check content 1`] = `
"{\\"$ctxType\\":\\"group\\",\\"$ctxId\\":321,\\"userId\\":123,\\"message\\":\\"foo\\"}
{\\"$ctxType\\":\\"group\\",\\"$ctxId\\":321,\\"userId\\":789,\\"message\\":\\"baz\\"}
"
`;
